rclcpp/src/rclcpp/node_builtin_executor.cpp

```diff
@@ -0,0 +1,188 @@
+      node_logging,
+      node_options
+))
+{}
+
+NodeBuiltinExecutor::~NodeBuiltinExecutor()
+{}
+
+NodeBuiltinExecutor::NodeBuiltinExecutorImpl::NodeBuiltinExecutorImpl(
+  node_interfaces::NodeBaseInterface::SharedPtr node_base,
+  node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
+  node_interfaces::NodeServicesInterface::SharedPtr node_services,
+  node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
+  const NodeOptions & node_options)
+: node_base_(node_base),
+  node_topics_(node_topics),
+  node_services_(node_services),
+  node_logging_(node_logging)
+{
+  if (node_options.enable_logger_service()) {
+    add_logger_services();
+  }
+
+  if (builtin_callback_group_) {
+    ExecutorOptions executor_option;
+    executor_option.context = node_options.context();
+
+    executor_ = std::make_shared<executors::SingleThreadedExecutor>(executor_option);
+    executor_->add_callback_group(
+      builtin_callback_group_,
+      node_base_);
+
+    executor_promise_ = std::promise<void>{};
+    thread_ = std::thread(
+      [this]() {
+        auto future = executor_promise_.get_future();
+        executor_->spin_until_future_complete(future);
+      }
+    );
+  }
+}
+
+NodeBuiltinExecutor::NodeBuiltinExecutorImpl::~NodeBuiltinExecutorImpl()
+{
+  if (thread_.joinable()) {
+    executor_promise_.set_value();
+    executor_->cancel();
+    thread_.join();
+  }
+}
+
+void
+NodeBuiltinExecutor::NodeBuiltinExecutorImpl::add_logger_services()
+{
+  const rclcpp::QoS & qos_profile = rclcpp::ServicesQoS();
+  const std::string node_name = node_base_->get_name();
+  auto callback_group = get_callback_group();
+
+  get_loggers_service_ = rclcpp::create_service<rcl_interfaces::srv::GetLoggerLevels>(
+    node_base_, node_services_,
+    node_name + "/get_logger_levels",
+    [](
+      const std::shared_ptr<rmw_request_id_t>,
+      const std::shared_ptr<rcl_interfaces::srv::GetLoggerLevels::Request> request,
+      std::shared_ptr<rcl_interfaces::srv::GetLoggerLevels::Response> response)
+    {
+      int ret = 0;
+      for (auto & n : request->names) {
+        rcl_interfaces::msg::LoggerLevel level;
+        level.name = n;
+        ret = rcutils_logging_get_logger_level(n.c_str());
+        if (ret < 0) {
+          level.level = 0;
+        } else {
+          level.level = (uint8_t)ret;
+        }
+        response->levels.push_back(std::move(level));
+      }
+    },
+    qos_profile, callback_group);
+
+  set_loggers_service_ = rclcpp::create_service<rcl_interfaces::srv::SetLoggerLevels>(
+    node_base_, node_services_,
+    node_name + "/set_logger_levels",
+    [](
+      const std::shared_ptr<rmw_request_id_t>,
+      const std::shared_ptr<rcl_interfaces::srv::SetLoggerLevels::Request> request,
+      std::shared_ptr<rcl_interfaces::srv::SetLoggerLevels::Response> response)
+    {
+      int ret = 0;
+      auto result = rcl_interfaces::msg::SetLoggerLevelsResult();
+      for (auto & l : request->levels) {
```

Comment:
```suggestion
      for (auto & level : request->levels) {
```

Reply:
Okay

Reply:
IMO, the meaning of `levels` in https://github.com/ros2/rcl_interfaces/pull/154 is vague.
Because it seems like a level array, the level might be treated as the `uint8 level` in the LoggerLevel.